src/wifi/model/periodic-dmg-wifi-scheduler.cc

```diff
@@ -0,0 +1,539 @@
+    }
+}
+
+uint32_t
+PeriodicDmgWifiScheduler::GetAllocationDuration (uint32_t minAllocation, uint32_t maxAllocation)
+{
+  NS_LOG_FUNCTION (this << minAllocation << maxAllocation);
+  return maxAllocation;
+}
+
+StatusCode
+PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecElement &dmgTspec, const DmgAllocationInfo &info)
+{
+  NS_LOG_FUNCTION (this);
+
+  StatusCode status;
+  uint32_t allocDuration;
+
+  if (m_availableSlots.begin () == m_availableSlots.end ())
+    {
+      NS_LOG_DEBUG ("There are no free available slots in the DTI.");
+      status.SetFailure ();
+      return status;
+    }
+
+  if (info.GetAllocationFormat () == ISOCHRONOUS)
+    {
+      allocDuration = GetAllocationDuration (dmgTspec.GetMinimumAllocation (), dmgTspec.GetMaximumAllocation ());
+      if (allocDuration < dmgTspec.GetMinimumAllocation ())
+        {
+          NS_LOG_DEBUG ("Unable to guarantee minimum duration.");
+          status.SetFailure ();
+          return status;
+        }
+    }
+  else if (info.GetAllocationFormat () == ASYNCHRONOUS)
+    {
+      // for asynchronous allocations, the Maximum Allocation field is reserved (IEEE 802.11ad 8.4.2.136)
+      allocDuration = dmgTspec.GetMinimumAllocation ();
+    }
+  else
+    {
+      NS_FATAL_ERROR ("Allocation Format not supported");
+    }
+
+  // Implementation of an admission policy for newly received requests. 
+  uint16_t allocPeriod = dmgTspec.GetAllocationPeriod ();
+  if (allocPeriod != 0)
+    {
+      // Proceed with the allocation of periodic SPs 
+      bool isMultiple = dmgTspec.IsAllocationPeriodMultipleBI ();
+      if (isMultiple)
+        {
+          NS_FATAL_ERROR ("Multiple BI periodicity is not supported yet.");
+        }
+
+      // spInterval is going to be passed to AddAllocationPeriod to specify the 
+      // distance between consecutive periodic SPs
+      uint32_t spInterval = uint32_t (m_biDuration.GetMicroSeconds () / allocPeriod);
+
+      NS_LOG_DEBUG ("Allocation Period " << uint32_t (allocPeriod) << " AllocDuration " << allocDuration << " Multiple " << isMultiple);
+      NS_LOG_DEBUG ("Schedule one SP every " << spInterval);
+
+      uint32_t startPeriodicAllocation = m_availableSlots[0].first;
+      uint32_t startFirstAllocation = startPeriodicAllocation;
+      uint32_t endAlloc;
+      uint8_t counter = 0;
+
+      uint8_t blocks = GetAvailableBlocks (allocDuration, spInterval);
+
+      if (blocks > 1)
+        {
+          while (counter < blocks)
+            {
+              NS_LOG_DEBUG ("Reserve from " << startPeriodicAllocation << " for " << allocDuration);
+
+              endAlloc = startPeriodicAllocation + allocDuration + m_guardTime;
+              UpdateAvailableSlots (startPeriodicAllocation, endAlloc);
+
+              startPeriodicAllocation += spInterval;
+              counter++;
+
+              // update remaining DTI time for consistency
+              m_remainingDtiTime -= (allocDuration + m_guardTime);
```

Comment:
can't this be part of `UpdateAvailableSlots`? there's a strong risk of forgetting to update it otherwise